osquery/tables/networking/windows/windows_firewall_rules.cpp

```diff
@@ -0,0 +1,203 @@
+/**
+ * Copyright (c) 2014-present, The osquery authors
+ *
+ * This source code is licensed as defined by the LICENSE file found in the
+ * root directory of this source tree.
+ *
+ * SPDX-License-Identifier: (Apache-2.0 OR GPL-2.0-only)
+ */
+
+#include <string>
+
+#include <atlbase.h>
+#include <netfw.h>
+#include <windows.h>
+
+#include <osquery/core/core.h>
+#include <osquery/core/tables.h>
+#include <osquery/logger/logger.h>
+#include <osquery/utils/conversions/windows/strings.h>
+
+namespace osquery {
+namespace tables {
+
+HRESULT populateRow(INetFwRule* rule, Row& r);
+
+QueryData genFirewallRules(QueryContext& context) {
+  QueryData results;
+
+  CComPtr<INetFwPolicy2> netFwPolicy;
+  HRESULT hr = netFwPolicy.CoCreateInstance(CLSID_NetFwPolicy2);
+  if (FAILED(hr)) {
+    TLOG << "Failed to instantiate INetFwPolicy2";
+    return results;
+  }
+
+  CComPtr<INetFwRules> rules;
+  hr = netFwPolicy->get_Rules(&rules);
+  if (FAILED(hr)) {
+    TLOG << "Failed to get firewall rules";
+    return results;
+  }
+
+  CComPtr<IUnknown> enumerator;
+  hr = rules->get__NewEnum(&enumerator);
+  if (FAILED(hr)) {
+    TLOG << "Failed to get firewall rules enumerator";
+    return results;
+  }
+
+  CComPtr<IEnumVARIANT> enumvar;
+  hr = enumerator->QueryInterface(IID_PPV_ARGS(&enumvar));
+  if (FAILED(hr)) {
+    TLOG << "Failed to get firewall rules enumerator interface";
+    return results;
+  }
+
+  do {
+    CComVariant value;
+    ULONG fetched = 0;
+    hr = enumvar->Next(1, &value, &fetched);
+    if (FAILED(hr)) {
+      TLOG << "Failed to enumerate next firewall rule";
+      break;
+    }
+
+    if (hr != S_FALSE) {
+      hr = value.ChangeType(VT_DISPATCH);
+      if (FAILED(hr)) {
+        TLOG << "Failed to get IDispatch for enumerated rule";
+        break;
+      }
+
+      CComPtr<INetFwRule> rule;
+      hr = (V_DISPATCH(&value))->QueryInterface(IID_PPV_ARGS(&rule));
+      if (FAILED(hr)) {
+        TLOG << "Failed to get firewall rule interface";
+        break;
+      }
+
+      Row r;
+      hr = populateRow(rule, r);
+      if (FAILED(hr)) {
+        TLOG << "Failed to convert firewall rule to row";
+        break;
+      }
+
+      results.push_back(std::move(r));
+    }
+  } while (SUCCEEDED(hr) && hr != S_FALSE);
+
+  return results;
+}
+
+HRESULT populateRow(INetFwRule* rule, Row& r) {
```

Comment:
This function should receive a list of structures from `genFirewallRules` and render them to `QueryResults`. This makes it easy to add tests specifically targeting rendering

Reply:
That would be less efficient, would take more memory and more copying data around, but sure can change.

Reply:
> That would be less efficient, would take more memory and more copying data around, but sure can change.

This approach is preferred because it allows for better testing. Some tables return too much data for this approach, so we usually skip it there (example: tables accessing system logs)

Reply:
I found it hard to find any table implemented with the suggested approach in the current osquery code, vast majority do not follow this pattern, so this makes it very inconsistent. 
Currently on my development pristine Windows 10 VM there are 470 firewall rules.
Anyways, will update this PR shortly as you requested, just to get this change in.